prow/plugins/invalidcommitmsg/invalidcommitmsg.go

```diff
@@ -0,0 +1,174 @@
+	if err != nil {
+		return err
+	}
+	return handle(pc.GitHubClient, pc.Logger, pr, cp)
+}
+
+func handle(gc githubClient, log *logrus.Entry, pr github.PullRequestEvent, cp commentPruner) error {
+	// Only consider actions indicating that the code diffs may have changed.
+	if !hasPRChanged(pr) {
+		return nil
+	}
+
+	shouldComment := pr.Action == github.PullRequestActionSynchronize ||
+		pr.Action == github.PullRequestActionOpened
+
+	var (
+		org    = pr.Repo.Owner.Login
+		repo   = pr.Repo.Name
+		number = pr.Number
+	)
+
+	labels, err := gc.GetIssueLabels(org, repo, number)
+	if err != nil {
+		return err
+	}
+	hasInvalidCommitMsgLabel := github.HasLabel(invalidCommitMsgLabel, labels)
+
+	allCommits, err := gc.ListPRCommits(org, repo, number)
+	if err != nil {
+		return fmt.Errorf("error listing commits for pull request: %v", err)
+	}
+	log.Debugf("Found %d commits in PR", len(allCommits))
+
+	var invalidCommits []github.GitCommit
+	for _, commit := range allCommits {
+		if closeIssueRegex.MatchString(commit.Commit.Message) || atMentionRegex.MatchString(commit.Commit.Message) {
+			c := commit.Commit
+			c.SHA = commit.SHA
+			invalidCommits = append(invalidCommits, c)
+		}
+	}
+
+	// if we have the label but all commits are valid,
+	// remove the label and prune comments
+	if hasInvalidCommitMsgLabel && len(invalidCommits) == 0 {
+		if err := gc.RemoveLabel(org, repo, number, invalidCommitMsgLabel); err != nil {
+			log.WithError(err).Errorf("GitHub failed to remove the following label: %s", invalidCommitMsgLabel)
+		}
+		cp.PruneComments(shouldPrune(log))
+	}
+
+	// if we don't have the label and there are invalid commits,
+	// add the label
+	if !hasInvalidCommitMsgLabel && len(invalidCommits) != 0 {
+		if err := gc.AddLabel(org, repo, number, invalidCommitMsgLabel); err != nil {
+			log.WithError(err).Errorf("GitHub failed to add the following label: %s", invalidCommitMsgLabel)
+		}
+	}
+
+	// if there are invalid commits, add a comment
+	if len(invalidCommits) != 0 && shouldComment {
+		// prune old comments before adding a new one
+		cp.PruneComments(shouldPrune(log))
+
+		log.Debugf("Commenting on PR to advise users of invalid commit messages")
+		if err := gc.CreateComment(org, repo, number, fmt.Sprintf(commentBody, dco.MarkdownSHAList(org, repo, invalidCommits), plugins.AboutThisBot)); err != nil {
+			log.WithError(err).Warning("Could not create comment for invalid commit messages")
+		}
+	}
+
+	return nil
+}
+
+// hasPRChanged indicates that the code diff may have changed.
+func hasPRChanged(pr github.PullRequestEvent) bool {
+	switch pr.Action {
+	case github.PullRequestActionOpened:
+		return true
+	case github.PullRequestActionReopened:
+		return true
+	case github.PullRequestActionSynchronize:
+		return true
+	default:
+		return false
+	}
+}
+
+// shouldPrune finds comments left by this plugin.
+func shouldPrune(log *logrus.Entry) func(github.IssueComment) bool {
+	return func(comment github.IssueComment) bool {
```

Comment:
Neither `log` nor the outer function are needed. This anonymous func could be named `shouldPrune` and used directly.

Reply:
Done.